### PR TITLE
Make netty-codec 4.2 Maven artifact behave like 4.1 regarding its dependencies

### LINKED ISSUE #14371: Increased modularity of netty-codec is backward incompatible change for 4.1
Change made in #14241 to address #14239 is a breaking change for libraries that depend on any classes from protobuf/marshalling/compression packages of `netty-codec`.

As an example, it broke the pipeline we use in ServiceTalk to validate compatibility with the latest Netty 4.2 snapshots: https://github.com/apple/servicetalk/actions/runs/11032862432/job/30646115309

If we proceed with this change, the 4.2 release will be much harder to adopt and require intervention (build/dependencies manipulation) from application owners because libraries won't be able to make themselves compatible with both 4.1 and 4.2 netty releases at the same time.

Imho, this change should wait for 5.0.

### PR DESCRIPTION
**Pull request by vietj**
Motivation:

The modularization of `netty-codec` lead to stripping `compression`/`protobuf`/`marshalling` from this module resulting in breaking backward compatibility from a build dependency perspective.

Changes:

Several changes are necessary to accomplish this.

Introduce `netty-codec-base` artifact which holds the code of `netty-codec` declaring the `io.netty.codec` module.

`netty-codec-(compression|protobuf|marshalling)` artifacts depends on `netty-codec-base` instead of `netty-codec`.

`netty-codec` artifact now ships an empty jar declaring the `io.netty.codec.unused` module (which should be avoided) and depends on `netty-codec-(base|compression|marshalling|protobuf)` artifact, in addition it redeclares the non optional dependencies of marshalling and protobuf as optional.

Completed the modular guide with an explanation to the developer.

Required revapi changes.

Result:

Fixes #14371

### CODE REVIEW

**Comment by vietj** [comment]
@idelpivnitskiy do you mind having a look at this ?

File: codec-base/pom.xml [hunk 1]
```diff
@@ -0,0 +1,97 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  ~ Copyright 2012 The Netty Project
```
Comment (normanmaurer) [inline codec-base/pom.xml:3]:
```suggestion
  ~ Copyright 2024 The Netty Project
```

File: codec-base/pom.xml [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        <executions>
+          <execution>
+            <id>module-info</id>
+          </execution>
+        </executions>
+      </plugin>
+    </plugins>
+  </build>
+
+</project>
```
Comment (normanmaurer) [inline codec-base/pom.xml:96]: @vietj I think we also need to add this dependency to  `bom/pom.xml`

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: codec-base/pom.xml (hunk 2 above)
Comment (idelpivnitskiy) [reply codec-base/pom.xml:96]: +1

Also, how should we configure `netty-all` now? should it include `netty-codec` or 4 independent modules only?

Comment (normanmaurer) [reply codec-base/pom.xml:96]: I think all of it.

Comment (vietj) [reply codec-base/pom.xml:96]: thanks, I missed that point

File: testsuite-jpms/README.md [hunk 3]
```diff
@@ -126,7 +160,7 @@ Boot layer:
 - io.netty.buffer 
 - io.netty.codec.http 
 - java.logging 
-- io.netty.codec 
+- io.netty.codec.base
```
[the remark this answers is not in the document]
Comment (vietj) [reply testsuite-jpms/README.md:163]: yes that's a mistake I'll fix it

**Comment by vietj** [comment]
@normanmaurer @idelpivnitskiy PTAL

File: codec-base/pom.xml (hunk 2 above)
Comment (idelpivnitskiy) [reply codec-base/pom.xml:96]: Still need an entry in netty-all for netty-codec-base

Comment (vietj) [reply codec-base/pom.xml:96]: will do thanks

Comment (vietj) [reply codec-base/pom.xml:96]: done

**Review by idelpivnitskiy — APPROVED** [review]
Thank you!

**Comment by normanmaurer** [comment]
@vietj thanks !

**Pull request merged**